### PR TITLE
[c10/cuda] Reorganize device_count() and robustly surface ASAN warnings

### PR DESCRIPTION
**Pull request by dzhulgakov**
Summary:
Main change is to bring Caffe2's superior error messages for cuda initialization into c10 and use them in all code paths.

Basic logic:

| Case | Call to device_count() | init_cuda, e.g. allocating tensor |
| -- | -- | -- |
| all good | non-zero | just works |
| no gpus | 0, no warning | throw exception with good message |
| driver issues | 0, produce warning | throw exception with good message |
| out of memory with ASAN | 0, produce warning| throw exception with ASAN message |

Previously, the error thrown from init_cuda was very generic and the ASAN warning (if any) was buried in the logs.

Other clean up changes:
* cache device_count() always in a static variable
* move all asan macros in c10

Test Plan:
Hard to unittest because of build modes. Verified manually that the behavior from the table above holds by running the following script in different modes (ASAN/no-ASAN, CUDA_VISIBLE_DEVICES=):

```
print('before import')
import torch
print('after import')
print('devices: ', torch.cuda.device_count())
x = torch.tensor([1,2,3])
print('tensor creation')
x = x.cuda()
print('moved to cuda')
```

Differential Revision: D22824329

### CODE REVIEW

**Comment by dzhulgakov** [comment]
@smessmer @iotamudelta @ezyang  - do you know what magic tricks I need to do to make it work with HIP? I can see that `device_count` is special cased in the hipify script: https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py#L8069 so I don't get how it was working before :)

**Comment by ngimel** [comment]
cc @jeffdaily for hip questions. Windows failures are real.

**Comment by jeffdaily** [comment]
HIP failures are 
```
build/lib/libtorch_hip.so: undefined reference to `c10::hip::device_count()'
build/lib/libtorch_hip.so: undefined reference to `c10::hip::device_count_ensure_non_zero()'
```

File: c10/cuda/CMakeLists.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ configure_file(
 # and headers you add
 set(C10_CUDA_SRCS
     CUDAStream.cpp
+    CUDAFunctions.cpp
```
Comment (jeffdaily) [inline c10/cuda/CMakeLists.txt:24]: See note about about adding new source file to hipify mappings.

Comment (jeffdaily) [reply c10/cuda/CMakeLists.txt:24]: https://github.com/pytorch/pytorch/blob/cc906ee6190699840048497e575aa27f2c5c63e2/c10/cuda/CMakeLists.txt#L19-L24

**Review by ngimel — APPROVED** [review]
Approving, modulo hip and windows failures.

**Comment by ngimel** [comment]
C10_CUDA_API macro is likely what you need for visibility issues.

File: c10/cuda/CMakeLists.txt (hunk 1 above)
Comment (dzhulgakov) [reply c10/cuda/CMakeLists.txt:24]: I've seen that, but I don't see any .cpp files listed in cuda_to_hip_mappings.py. Note that I'm adding a .cpp file for an existing CUDAFunctions.h file and it's already listed in cuda_to_hip_mappings.

Also there's some special casing for device_count() in it but I can't figure out how it works: https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py#L8069

Comment (dzhulgakov) [reply c10/cuda/CMakeLists.txt:24]: Seems like it was the same missing C10_CUDA_API macro problem

**Pull request closed without merging**